### ISSUE TITLE
Add tests for optimizers and tools

### ISSUE THREAD

**Issue by samgdotson** [arfc/cairo#60]
This issue can be closed when all of the functions in optimizers and tools have been fully tested with tests in ``/tests/test_tools.py`` or ``/tests/test_optimizer.py``.

**Comment by nsryan2** [comment]
@samgdotson are tests supposed to end with a return? Some of the existing tests for tools and optimizers have returns at the end and others don't.

**Comment by samgdotson** [comment]
I think the answer is yes, they should have return statements. In general, all functions should have return statements. It is a bit different here since we will never be *calling* these functions, but I think for the sake of readability they should be included.

**Comment by nsryan2** [comment]
Awesome! Not at all related, when I run pytest I get that the test `test_esn_prediction_diffsize()` fails because of an AssertionError rather than an IndexError as the test says it should. I haven't changed anything about the test other than the name of parameters that it's using so I could make test for the save feature in the function. Should the tests for window size and mismatched inputs have different parameters, or is that a non-issue here?

**Comment by nsryan2** [comment]
Aditionally, for the function `optimal_values`, I tried applying the case where the two entries have the exact same numbers. I'm not sure if that's a physically possible situation, but it seemed like a good boundary to test. The function fails for arrays like :
`x = np.array([0, 1, 0]) y = np.array([1, 0, 0])`

I messed around with it and found that changing x_optimal and y_optimal to say:
`x_optimal = xset[int(index_min[0][0])]`
`y_optimal = yset[int(index_min[1][0])]`
in the function got rid of this issue and still yielded the same results for the example used in `test_optimal_values_pmone()` and I'm pretty sure it worked for the example I gave above as well. Is this a good idea?

**Comment by samgdotson** [comment]
Okay so there are two questions here: 
1) Should the tests for window size and mismatched inputs have different parameters? If you're writing a unit test, it should ideally test one thing. If the code is supposed to fail because of a mismatch in the size of data, then that should be the only thing that would fail (i.e. other parameters should be good). In that sense, yes, the two tests should have different parameters. Now, it looks like you're also using the same ``X_in`` in both tests. This is fine, but if ``X_in`` causes a failure, you should probably use the correct version ``X_in.T``. 
2) This is a good question about ``optimal_values`` it would be a good idea if the minimum value we're looking for was in either ``x_set`` or ``y_set``. But this is not the case. We are instead looking for a minimum _loss_ value. This loss matrix is not the outer product of ``x`` and ``y``, but rather it is generated by calculating the error of a prediction made from each combination of values in ``x_set`` and ``y_set``. I hope I'm being clear. For tests, the loss matrix should a random 3x3 matrix. E.g. ``loss = np.random.rand(3,3)``. If you want to test specific cases (like duplicate values) you can change specific values in that matrix.
One final thought on this, I would never expect two numbers to be exactly the same in a matrix like this. Although it could technically be possible due to a fluke of machine precision. It never hurts to have too many tests. If two values ARE identical, that would mean either A) there is something wrong with the code or B) they're both minima, in which case it doesn't matter as long as the function picks one of them out. Remember: the loss matrix has entries that are the error for a certain set of parameters, if they're equally good, it doesn't really matter which one we use. 

I hope this wasn't too much information. Keep up the good work
